lib/Mir.ml

```diff
@@ -145,18 +145,23 @@ type ('e, 's) prog =
   ; prog_path: string }
 [@@deriving sexp]
 
-type expr_typed_located =
-  { texpr_type: unsizedtype
-  ; texpr_loc: location_span sexp_opaque [@compare.ignore]
-  ; texpr: expr_typed_located expr
-  ; texpr_adlevel: autodifftype }
-[@@deriving sexp, hash, map, of_sexp]
-
-type stmt_loc =
-  { sloc: location_span sexp_opaque [@compare.ignore]
-  ; stmt: (expr_typed_located, stmt_loc) statement }
-[@@deriving hash, map, sexp, of_sexp]
+type 'm expr_with = {expr: 'm expr_with expr; emeta: 'm}
 
+and typed_and_located =
+  { expr_type: unsizedtype
+  ; expr_loc: location_span sexp_opaque [@compare.ignore]
+  ; expr_adlevel: autodifftype }
+
+and expr_typed_located = typed_and_located expr_with [@@deriving sexp]
+
+type ('e, 'm) stmt_with = {stmt: ('e, ('e, 'm) stmt_with) statement; smeta: 'm}
```

Comment:
Is it useful to be able to swap out the whole expression type, or do we only need to swap out the expression metadata type? If the latter, we could write something like
`
type ('e, 'm) stmt_with = {stmt: ('e expr_with, ('e, 'm) stmt_with) statement; smeta: 'm}
`
This was `'e` is expression metadata and `'m` is statement metadata.

Reply:
I thiiiink either one works? I guess that might help with some of the error messages.